pkg/virt-launcher/virtwrap/live-migration-source.go

```diff
@@ -413,13 +419,130 @@ func newMigrationMonitor(vmi *v1.VirtualMachineInstance, l *LibvirtDomainManager
 		options:                  options,
 		migrationErr:             migrationErr,
 		progressWatermark:        int64(0),
+		remainingData:            int64(0),
 		progressTimeout:          options.ProgressTimeout,
 		acceptableCompletionTime: options.CompletionTimeoutPerGiB * getVMIMigrationDataSize(vmi),
 	}
 
 	return monitor
 }
 
+func (m *migrationMonitor) shouldTriggerTimeout(elapsed int64, domSpec *api.DomainSpec) bool {
+	if m.acceptableCompletionTime == 0 {
+		return false
+	}
+
+	if domSpec.Metadata.KubeVirt.Migration != nil && domSpec.Metadata.KubeVirt.Migration.Mode == v1.MigrationPostCopy {
+		return false
+	}
+
+	if elapsed > m.acceptableCompletionTime {
+		return true
+	}
+
+	return false
+}
+
+func (m *migrationMonitor) shouldTriggerPostCopy(elapsed int64, domSpec *api.DomainSpec) bool {
+	if m.shouldTriggerTimeout(elapsed, domSpec) && m.options.AllowPostCopy {
+
+		return true
+	}
+	return false
+}
+
+func (m *migrationMonitor) isMigrationProgressing(domainSpec *api.DomainSpec) bool {
+	logger := log.Log.Object(m.vmi)
+
+	now := time.Now().UTC().Unix()
+
+	// If the migration is in post copy mode we should
+	// not abort the job as the migration would lose state
+	// so post copy can not timeout
+	if domainSpec.Metadata.KubeVirt.Migration != nil && domainSpec.Metadata.KubeVirt.Migration.Mode == v1.MigrationPostCopy {
+		return true
+	}
+
+	// check if the migration is progressing
+	progressDelay := now - m.lastProgressUpdate
+	if m.progressTimeout != 0 && progressDelay > m.progressTimeout {
+		logger.Warningf("Live migration stuck for %d sec", progressDelay)
+		return false
+	}
+
+	return true
+}
+
+func (m *migrationMonitor) processInflightMigration(dom cli.VirDomain) *inflightMigrationAborted {
+	logger := log.Log.Object(m.vmi)
+
+	// Migration is running
+	now := time.Now().UTC().Unix()
+	elapsed := now - m.start
+
+	if (m.progressWatermark == 0) ||
+		(m.progressWatermark > m.remainingData) {
+		m.progressWatermark = m.remainingData
+		m.lastProgressUpdate = now
+	}
+
+	domainSpec, err := m.l.getDomainSpec(dom)
+	if err != nil {
+		logger.Reason(err).Error("failed to get domain spec info")
```

Comment:
> I see this repeating here several times, sometimes even without breaking out early. Very strange and unclear.

my guess is that no one is confident enough to break out of the loop and fail the migration due to an error like this... likely because no one can enumerate all the reasons this error can occur in order to know what the appropriate actions is. the conservative action is to keep retrying.